File: solutions/lua/prime-factors/1/prime-factors.lua
--[[
Primes smaller than n are numbers up to sqrt(n).

We could use a prime generator to capture all the primes smaller than n
and then test to see which one are divisors of n but this will be slow in lua
for large value of n because of the prime array construction and manipulation.

If instead we try all numbers between 2 and sqrt(n) against n or its remainder (once divided
by previously found prime factors) we will get the same result. All divisions by non-prime will fails
to be exact division as the remainder at this stage has already been divided by the factors of the
non-prime numbers.

We can further optimize by skipping all even numbers greater than 2
]]
return function(n)

    local factors = {}

    -- find how many times 2 is a prime factor
    while n % 2 == 0 do
        table.insert(factors, 2)
        n = n / 2
    end

    -- for each odd number up to sqrt(n) find how many times they are prime factors
    for i = 3, math.sqrt(n), 2 do
        while n % i == 0 do
            table.insert(factors, i)
            n = n / i
        end
    end

    -- if there is any remainder, it is itself a prime factor
    if n > 2 then
        table.insert(factors, n)
    end

    return factors
end

File: solutions/javascript/prime-factors/1/prime-factors.js
// Returns the list of primes smaller or equal to n.
function primes(n) {

  if (n < 2) return [];

  // Accumulate the list of known primes in 'primes'.
  // We know 2 is the only even number that is prime.
  // Go through the odd numbers between 3 and n and test if they
  // are divisible by any of the already found primes. If they are
  // not, then add to the list of know primes.
  
  const primes = [2];

  for (let candidate = 3; candidate <= n; candidate += 2) {
    let divisorFound = false;
    for (const prime of primes) {
      if (candidate % prime == 0) {
        divisorFound = true;
        break;
      }
    }
    if (!divisorFound) primes.push(candidate);
  }
  return primes;
}

// Returns the list of primeFactors for n.
export function primeFactors(n) {

  // All prime factors must be lower than sqrt(n) (except for the last one)
  const possibleFactors = primes(Math.trunc(Math.sqrt(n)))
  let factors = [];
  for (const f of possibleFactors) {
    while (n % f == 0) {
      factors.push(f)
      n /= f
    }
  }
  // What ever remain is the last prime factor.
  if (n > 1) factors.push(n);
  return factors;
}